### PR TITLE
test: de-mock the view tests (PR A) — real KeaClient + HTTP-boundary stub

### PR DESCRIPTION
**Pull request by marcinpsk**
**Draft / WIP — building incrementally.** First step of the codex-ratified plan (PR A → C → B).

## Why
The view tests patch `netbox_kea.models.KeaClient` with a `MagicMock` and assert call-args, so `command()`'s real payload is never built or inspected — a payload regression (e.g. a stray/missing `service` key) stays green.

## This step
- **`netbox_kea/tests/kea_stub.py`** — `stub_kea({command: response})` patches `requests.Session.post` at the class level (covers `KeaClient.clone()` on worker threads) and records the request bodies. Real view → real `get_client()` → real `command()` → stubbed HTTP.
- Converted `test_post_htmx_single_lease_returns_hx_refresh` to assert the real `lease4-del` payload (`arguments` + `service`) — proving the pattern before scaling.

## Next
Convert the rest of `test_views_leases.py` in flow-grouped batches (search paths, reserved-badge via real `reservation_get`, pagination), then the other view-test files. This is the safety net that makes PR C (drop `service`, endpoint-aware) verifiable.

## Summary by CodeRabbit

* **Tests**
  * Added a command-aware Kea HTTP stub for view/API tests, including request/endpoint recording, per-command canned responses, queued/paginated sequencing, callable-driven payloads, and thread-safe handling across concurrent requests.
  * Added a dedicated test suite to validate response normalization, dispatch/registration errors, exception propagation, and concurrency guarantees.
  * Refactored lease, reservation, server, subnet, options, DHCP control, sync, and combined view/API tests to use the HTTP-boundary stubbing approach, asserting generated commands/payloads plus pagination/cursor, redirects, and robust error/edge-case behavior.

### CODE REVIEW

File: netbox_kea/tests/kea_stub.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        if cmd not in self._responses:
+            raise AssertionError(f"KeaHttpStub: no response registered for command {cmd!r} (url={url})")
+        spec = self._responses[cmd]
+        if callable(spec):
+            payload = spec(body)
+        elif isinstance(spec, list):
+            payload = spec.pop(0) if len(spec) > 1 else spec[0]
+        else:
+            payload = spec
+        return _http_response(payload if isinstance(payload, list) else [payload])
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/kea_stub.py]: > **Make queued responses explicit and synchronize their dispatch.**
> Line 53 treats every `list` as a FIFO queue, so a valid multi-entry Kea response such as `[{"result": 0}, {"result": 3}]` is truncated to one entry; `[]` also raises `IndexError`. Additionally, `len(spec)` plus `pop(0)` is non-atomic despite the worker-thread support claim.

Fixed in 03ecfa0.

- A plain `list` value is now returned **verbatim** as the multi-service `.json()` payload — no truncation, and `[]` is a valid empty payload (no more `IndexError`).
- Sequential responses use an explicit `queued()` / `ResponseQueue` (backed by a `deque`); all pagination / partial-failure call sites were migrated to it.
- Request recording and queue dispatch are guarded by a `threading.Lock`, so the non-atomic length-check-then-pop can no longer race across the cloned worker-thread clients.
- Added `netbox_kea/tests/test_kea_stub.py` covering verbatim multi-entry payloads, `queued()` sequencing (last repeats), the empty-queue error, and concurrent dispatch (each queued entry consumed exactly once — this would flake without the lock).

File: netbox_kea/tests/test_views_leases.py [hunk 2]
```diff
@@ -163,56 +165,56 @@ class TestReservedBadgeOnLeases(_ViewTestBase):
...
+                "config-get": self._CONFIG4,
+                "lease4-get": {"result": 0, "arguments": {"ip-address": "192.168.1.100", **self._LEASE4}},
+                "reservation-get": {"result": 3},  # not found
+            }
+        ):
+            response = self._htmx_get(url, {"by": "ip", "q": "192.168.1.100"})
 
         self.assertEqual(response.status_code, 200)
         # The column header says "Reserved" — check no badge link is rendered
         self.assertNotContains(response, 'text-decoration-none">Reserved</a>')
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_views_leases.py:208]: > **Assert successful HTMX rendering in both negative cases.**
> A caught lookup failure renders `exception_htmx.html` with HTTP 200 and no badge, so these tests can pass without exercising the intended graceful table rendering. Assert a success-only response signal and lease content.

Fixed in 03ecfa0. Both negative reserved-badge tests (`test_no_reserved_badge_when_no_reservation`, `test_no_crash_when_host_cmds_unavailable`) now assert `HX-Push-Url` in the response headers — set only on the success render path (`leases.py` `response["HX-Push-Url"] = stripped_return_url`), not on the exception partial — plus `assertContains(response, "192.168.1.100")` for the lease row. A fallback to `exception_htmx.html` can no longer pass them silently.

File: netbox_kea/tests/test_views_leases.py [hunk 3]
```diff
@@ -1900,16 +1848,18 @@ def test_lease_journal_db_error(self):
...
-        )
+    def test_htmx_exception_returns_error_partial(self):
+        # A RuntimeError from the lease fetch must be caught and rendered as the HTMX error partial.
+        with stub_kea({"config-get": self._CONFIG4, "lease4-get": RuntimeError("boom")}):
+            response = self.client.get(
+                self._url() + "?q=10.0.0.1&by=ip",
+                HTTP_HX_REQUEST="true",
+            )
         # Must not crash — returns HTMX error partial
         self.assertIn(response.status_code, (200, 500))
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_views_leases.py]: > Tighten the status assertion so an uncaught exception can't pass silently.

Fixed in `73f2bee`: the assertion is now `assertEqual(response.status_code, 200)` (was `assertIn(status, (200, 500))`). Confirmed the HTMX handler catches the lease-fetch `RuntimeError` and renders the error partial at 200, so the tightened assertion passes and now catches a regression where the handler stops catching.

File: netbox_kea/tests/test_views_sync.py [hunk 4]
```diff
@@ -241,47 +279,51 @@ def test_sync_raises_not_implemented(self):
...
-            mock_sync.assert_not_called()
-
-    @patch("netbox_kea.models.KeaClient")
-    def test_falls_back_on_kea_exception(self, MockKeaClient):
-        """When reservation_get_by_ip raises KeaException, response is 400."""
-        from netbox_kea.kea import KeaException
-
-        MockKeaClient.return_value.reservation_get_by_ip.side_effect = KeaException({"result": 1, "text": "not found"})
+        self.assertEqual(response.status_code, 400)
+        mock_sync.assert_not_called()
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_views_sync.py]: > Stale test name/docstring contradicts the asserted behavior.

Fixed in `73f2bee`: renamed `test_falls_back_to_synthetic_when_reservation_not_found` → `test_reservation_not_found_returns_400_without_sync` (the test asserts 400 + `_sync` not called, i.e. no synthetic fallback). Setup/assertions unchanged.

File: netbox_kea/tests/test_reservation_views.py [hunk 5]
```diff
@@ -46,25 +47,160 @@
...
+    """
+    return {
+        "result": 0,
+        "arguments": {f"subnet{version}": [{"id": subnet_id, "pools": [{"pool": p} for p in (pools or [])]}]},
+    }
+
+
+def _res_get(reservation):
+    """A ``reservation-get`` result: the host fields Kea returns directly inside ``arguments``."""
+    return {"result": 0, "arguments": dict(reservation)}
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_reservation_views.py]: > Consolidate the shared reservation stub helpers into `kea_stub.py`.

Fixed in `73f2bee`: moved `_res_page` / `_res_get` / `_subnet_get` next to `stub_kea`/`queued` in `kea_stub.py` and removed the duplicate local defs across `test_reservation_views`, `test_views_reservations`, `test_views_sync`, `test_global_views`, `test_jobs`, and `test_sync_views` (the duplication was wider than the two flagged files), importing them from `kea_stub` (matching the existing `_make_db_server`-from-`utils` convention). Added a shape-locking unit test in `test_kea_stub.py`. The trivial `{"result": 3}` sentinel constants (`_RES_EMPTY_PAGE`/`_RES_NOT_FOUND`/`_LEASE_NOT_FOUND`) were left file-local — they carry no shape to drift.

File: netbox_kea/tests/test_sync_views.py [hunk 6]
```diff
@@ -424,24 +414,12 @@ class TestReservation6BulkSyncView(_SyncViewBase):
...
-            ],
-            0,
-            0,
-        )
-        self.client.post(self._url())
-        ip = NbIP.objects.filter(address__startswith="2001:db8::1").first()
+        hosts = [{"subnet-id": 1, "duid": "00:01:aa:bb", "ip-addresses": ["2001:db8::1"], "hostname": "host-v6"}]
+        with stub_kea({"reservation-get-page": _res_page(hosts)}):
+            self.client.post(self._url())
+        ip = NbIP.objects.filter(address__startswith="2001:db8::1/").first()
```
Comment (marcinpsk) [inline netbox_kea/tests/test_sync_views.py:417]: @coderabbitai — addressed the outside-diff finding:
- `netbox_kea/tests/test_sync_views.py` (v6 bulk-sync `NbIP` lookup in `test_post_bulk_syncs_v6_reservations`) → fixed in `73f2bee`: anchored with a trailing slash (`address__startswith="2001:db8::1/"`) to match the other IP assertions and prevent a bare prefix matching adjacent addresses.

File: netbox_kea/tests/test_api_views_direct.py [hunk 7]
```diff
@@ -347,107 +372,99 @@ def test_generic_exception_returns_500(self):
...
-        server.get_client.return_value = mock_client
-        response = view._reservation_search(_make_request({"subnet_id": "1"}), version=4)
+        pages = queued(
+            self._page([host1], next_from=1, next_source=1),  # not exhausted
+            self._page([host2]),  # exhausted
+        )
+        with stub_kea({"reservation-get-page": pages}):
+            response = view._reservation_search(_make_request({"subnet_id": "1"}), version=4)
         self.assertEqual(response.status_code, status.HTTP_200_OK)
         self.assertEqual(response.data["count"], 2)
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_api_views_direct.py]: > Reuse `_res_page` from `netbox_kea/tests/kea_stub.py` instead of re-implementing the pagination payload shape.

Fixed in `c6c98c7`: `test_api_views_direct.py` now imports `_res_page` and drops the local `_page` method (all 3 call sites use `_res_page(...)`); `test_api_reservations.py` builds `_RESERVATION4_PAGE_RESPONSE` and the inline `page1`/`page2` via `_res_page(...)`, and the unused `_RESERVATION6_PAGE_RESPONSE` duplicate page fixture was removed. codex (gpt-5.6-sol, xhigh) reviewed the diff and confirmed the generated payloads are behaviorally equivalent.

File: netbox_kea/tests/test_views_sync.py [hunk 8]
```diff
@@ -1,39 +1,74 @@
...
+            "valid-lft": 86400,
+            "cltt": 1700000000,
+            "subnet-id": 1,
+        },
+    }
+
+
+def _subnet_list(version, subnets):
+    """A ``subnet{v}-list`` payload (used by ``reservation_get_by_ip`` to find candidate subnets)."""
+    return {"result": 0, "arguments": {"subnets": list(subnets)}}
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_views_sync.py]: > Promote `_subnet_list` to `kea_stub`.

Fixed in `c6c98c7`: added a version-parameterized `_subnet_list(version, subnets)` (list-of-subnet-dicts form) to `kea_stub.py` next to `_subnet_get`, removed the local copies in `test_views_sync.py` and `test_sync_views.py`, and imported the shared builder. `test_sync_views.py`'s two call sites now pass the shared list form (`[{"id": 1, "subnet": <cidr>}]`). Added `_subnet_list` to the `kea_stub` shape-lock test.

File: netbox_kea/tests/test_views_sync.py [hunk 9]
```diff
@@ -241,47 +260,51 @@ def test_sync_raises_not_implemented(self):
...
-    def test_falls_back_on_kea_exception(self, MockKeaClient):
-        """When reservation_get_by_ip raises KeaException, response is 400."""
-        from netbox_kea.kea import KeaException
-
-        MockKeaClient.return_value.reservation_get_by_ip.side_effect = KeaException({"result": 1, "text": "not found"})
+        self.assertEqual(response.status_code, 400)
+        mock_sync.assert_not_called()
 
-        with patch("netbox_kea.views.ServerReservation4SyncView._sync") as mock_sync:
+    def test_kea_exception_returns_400_without_sync(self):
```
Comment (marcinpsk) [inline netbox_kea/tests/test_views_sync.py:299]: @coderabbitai — addressed the outside-diff finding:
- `netbox_kea/tests/test_views_sync.py` (the two remaining stale `falls_back` names) → fixed in `c6c98c7`: `test_falls_back_on_kea_exception` → `test_kea_exception_returns_400_without_sync` and `test_falls_back_on_request_exception` → `test_request_exception_returns_400_without_sync` (both assert 400 + `_sync` not called, i.e. no synthetic fallback). Assertions unchanged.

File: netbox_kea/tests/test_views_sync.py [hunk 10]
```diff
@@ -291,119 +314,94 @@ class TestReservation6SyncViewFetchLiveData(_ViewTestBase):
...
-
-        MockKeaClient.return_value.reservation_get_by_ip.side_effect = req.RequestException("timeout")
-
-        response = self.client.post(self._url(), {"ip_address": "2001:db8::1", "hostname": "fallback6"})
-
+    def test_request_exception_returns_400_without_sync(self):
+        """When the subnet-list call raises a transport error, response is 400 (no sync)."""
+        with stub_kea({"subnet6-list": requests.RequestException("timeout")}):
+            response = self.client.post(self._url(), {"ip_address": "2001:db8::1", "hostname": "fallback6"})
         self.assertEqual(response.status_code, 400)
```
[the remark this answers is not in the document]
Comment (marcinpsk) [reply netbox_kea/tests/test_views_sync.py:331]: > Assert that the IPv6 request-exception path skips `_sync`.

Fixed in `7ce9dba`: `test_request_exception_returns_400_without_sync` now patches `ServerReservation6SyncView._sync` and asserts `mock_sync.assert_not_called()` alongside the 400, mirroring the v4 sibling `test_kea_exception_returns_400_without_sync` — the transport error makes `_fetch_live_data` return None, so `_sync` is never reached, and this now guards that.
